### PR TITLE
maint: req py36+ and jh 1.5.1+, fix tests, add RELEASE.md, add pre-commit hooks, add dependabot

### LINKED ISSUE #274: Add dependabot to bump github actions monthly
We don't want to get spammed with PRs related to this, but we do want to get updated automatically.

### LINKED ISSUE #275: Add RELEASE.md documentation

### LINKED ISSUE #276: Refactor to not use: asyncio_generator, tornado.gen.coroutine, tornado.gen.sleep
We can refactor to not need use of the following any more:
- asyncio_generator
- tornado.gen.coroutine
- tornado.gen.sleep

Also, we can and need to refactor tests to use `event_loop` fixture and no longer anything about `io_loop`.

### PR DESCRIPTION
**Pull request by consideRatio**
This PR is making some progress towards general maintenance needs tracked in #270, doing things done for other repos in the jupyterhub org.

It is a large PR that combines various things that could be made into at least some separate PRs, but has for now been combined to avoid merge conflicts. If its too much work to review in this form or needs various updates, I'll try to extract a few PRs to make it easier to review going onwards.

- Addresses the jupyterhub part of #271
- Fixes #274
- Fixes #275
- Fixes #276

### CODE REVIEW

**Comment by consideRatio** [comment]
Ping also to @rkdarst who has contributed a lot to this repo!

File: batchspawner/api.py
Comment (ryanlovett) [inline batchspawner/api.py:4]: A [recommended](https://peps.python.org/pep-0008/#imports) order of imports is

standard library:
> `import json`

then related third part imports:
> `from tornado import web`

then local application/library specific imports:
> `from jupyterhub.apihandlers import APIHandler, default_handlers`

Although they also recommend putting a blank line in between each group, so one line before tornado. I suppose this change would be congruent with the recommendation if tornado is categorized as local application rather than related third party, but I think of it as more the latter.

**Review by ryanlovett — COMMENTED** [review]
Thanks for doing this @consideRatio !

I personally am okay with pushing minimum dependencies forward, but system-python users on older distributions may feel differently. I defer to @mbmilligan, who also worked on the test matrix some months back.

File: .github/workflows/test.yaml
Comment (consideRatio) [reply .github/workflows/test.yaml:33]: Ah, I found these notes:

image

---

For reference, Python 3.7 reaches end-of-life state 28th June.

image

File: batchspawner/api.py
Comment (consideRatio) [reply batchspawner/api.py:4]: The formatting is a result of autoformatting via pre-commit that has declared use of `isort`, configured with `--profile=black`.

image

I figure the behavior now is that `tornado` and `jupyterhub` are considered third party imports and grouped together.

I think [isort can be configured to associate specific imports in different ways](https://pycqa.github.io/isort/docs/configuration/options.html#known-local-folder), for example considering jupyterhub or tornado as something to be put in another category.

I'm very positively towards the use of autoformatting, but I have no big opinions on how.

Comment (consideRatio) [reply batchspawner/api.py:4]: I read about third party and first party imports in https://github.com/PyCQA/isort/issues/1057.

I think for example pytest, tornado, and jupyterhub make sense to see as third party imports, while `import batchspawner` would be a first party import, and `import .batchspawner` would be a local import.

File: pyproject.toml
Comment (consideRatio) [inline pyproject.toml:55]:
```suggestion
github_url = "https://github.com/jupyterhub/batchspawner"
```

**Comment by mbmilligan** [comment]
@consideRatio 
Hi Erik, just to note that this is much needed work and I'm really happy to see you taking it on! I'm on vacation through July 5 so I probably won't have time to dig into a review before then.

**Comment by consideRatio** [comment]
Hi @mbmilligan I hope you are well!

Do you think you will find time to review this PR and others to get us towards a release? I also wondered if you feel that its fine if someone else in the jupyterhub team reviewed and merged some PRs?

**Comment by mbmilligan** [comment]
I would *absolutely* welcome other folks familiar with the codebase helping
out with reviews and testing of PRs. I prefer to have a single person
actually pushing the merges, but I could do that quickly if given good
documentation that proposed changes are okay to merge.

**Review by yuvipanda — CHANGES_REQUESTED** [review]
As someone who doesn't run HPC systems nor has to maintain anything on CentOS (praise be!), I'm happy to merge this PR if python 3.6 removal is split out into a different PR. I would like someone from batchspawner / HPC community to make that choice as I don't feel qualified to make that.

Thank you so much for working on this, @consideRatio

**Comment by yuvipanda** [comment]
Upon re-reading @mbmilligan's comment in https://github.com/jupyterhub/batchspawner/pull/273#issuecomment-1725817717, let me amend to say that I'm very happy to approve all the changes except for 3.6 support dropping. My interpretation of the comment is that @mbmilligan would still prefer to be the one pushing the merge button so am happy to leave that up to them :)

**Comment by consideRatio** [comment]
Thanks for reviewing @yuvipanda and @mbmilligan!

I've pushed a commit reverting dropping support for python 3.6. For reference note that jupyterhub 3 requires python 3.7, so users with python 3.6 are stuck with jupyterhub 2 for now also.

I also pushed a code coverage related config omitting coverage of tests.

**Review by yuvipanda — APPROVED** [review]
Thank you, @consideRatio! This looks good to me.

**Comment by yuvipanda** [comment]
I've reviewed and approved this, @mbmilligan! I am not 100% sure if your comment meant you want to be the person hitting merge, so I'll just give this a few days before I hit merge :)

**Comment by yuvipanda** [comment]
@consideRatio I think you can also have the 3.6 drop as a separate PR! I personally think it's ok to drop that, but I'd like someone from the HPC community to make that choice. We can definitely 100% do so once CentOS EOLs in 2024.

**Comment by mahendrapaipuri** [comment]
We use batchspawner on our HPC platform and well, we dont really rely on python distribution shipped out of OS. We install all Jupyter related stack in a separate conda/mamba environment. Moreover JupyterHub 4+ supports only Python 3.8+, so, I do not really see a point on supporting Python 3.6+ for batchspawner.

Saying so, I agree with @yuvipanda that we should drop support for Python 3.6 in a separate PR. This is due to the fact that there is no release of batchspawner that supports JupyterHub 3.* (#277) which supports Python 3.7. So, I think it makes sense to make a release that is compatible with JupyterHub 3.* and then drop support for the next releases that are meant for JupyterHub 4.*.

**Comment by yuvipanda** [comment]
@mbmilligan are you ok with me merging this PR now?

**Comment by mbmilligan** [comment]
> @mbmilligan are you ok with me merging this PR now?

Thanks for the ping and for the work pushing this forward. Merging now.

**Comment by mbmilligan** [comment]
> We use batchspawner on our HPC platform and well, we dont really rely on python distribution shipped out of OS. We install all Jupyter related stack in a separate conda/mamba environment. Moreover JupyterHub 4+ supports only Python 3.8+, so, I do not really see a point on supporting Python 3.6+ for batchspawner.
> 
> Saying so, I agree with @yuvipanda that we should drop support for Python 3.6 in a separate PR. This is due to the fact that there is no release of batchspawner that supports JupyterHub 3.* (#277) which supports Python 3.7. So, I think it makes sense to make a release that is compatible with JupyterHub 3.* and then drop support for the next releases that are meant for JupyterHub 4.*.

That was my thinking as well. I'll merge this PR now and set up a release, and after that release we can drop additional support lines.

**Review by mbmilligan — APPROVED** [review]

**Comment by yuvipanda** [comment]
Thank you for the merge, @mbmilligan!

And thank you for doing all this work, @consideRatio!

**Pull request merged**
